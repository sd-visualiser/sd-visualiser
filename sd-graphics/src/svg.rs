use egui::{Color32, emath::TSTransform};
use sd_core::hypergraph::{
    generic::Ctx,
    traits::{WireType, WithType, WithWeight},
};
use svg::{
    Document, Node,
    node::element::{Circle, Group, Line, Path, Rectangle, Text, path::Data},
};

use crate::{
    common::SCALE,
    shape::{Shape, Shapes},
};

impl<T: Ctx> Shape<T>
where
    <<T as Ctx>::Edge as WithWeight>::Weight: WithType,
{
    pub(crate) fn to_svg(&self) -> Box<dyn Node> {
        match self {
            Self::Operation {
                center,
                radius,
                label,
                ..
            } => {
                let x_size =
                    f32::from(*radius) / 20.0 * f32::from(label.chars().count().max(1) as u16 + 1);
                Box::new(
                    Group::new()
                        .add(
                            Rectangle::new()
                                .set("x", center.x - x_size / 2.0)
                                .set("y", center.y - f32::from(*radius) / 20.0)
                                .set("width", x_size)
                                .set("height", f32::from(*radius) / 10.0)
                                .set("rx", f32::from(*radius) / 20.0)
                                .set("ry", f32::from(*radius) / 20.0)
                                .set("fill", "white")
                                .set("stroke", "black")
                                .set("stroke-width", 1),
                        )
                        .add(
                            Text::new(html_escape::encode_text(label))
                                .set("x", center.x)
                                .set("y", center.y)
                                .set("font-size", 16)
                                .set("font-family", "monospace")
                                .set("text-anchor", "middle")
                                .set("dominant-baseline", "middle"),
                        ),
                )
            }
            Self::CircleFilled {
                center,
                radius,
                addr,
                ..
            } => Box::new(
                Circle::new()
                    .set("cx", center.x)
                    .set("cy", center.y)
                    .set("r", f32::from(*radius) / 20.0)
                    .set(
                        "fill",
                        wire_type_to_svg_colour(addr.weight().get_type()).to_hex(),
                    ),
            ),
            Self::Rectangle { rect, .. } => Box::new(
                Rectangle::new()
                    .set("x", rect.min.x)
                    .set("y", rect.min.y)
                    .set("width", rect.width())
                    .set("height", rect.height())
                    .set("fill", "none")
                    .set("stroke", "gray")
                    .set("stroke-width", 1),
            ),
            Self::Line {
                start, end, addr, ..
            } => Box::new({
                Line::new()
                    .set("x1", start.x)
                    .set("y1", start.y)
                    .set("x2", end.x)
                    .set("y2", end.y)
                    .set(
                        "stroke",
                        wire_type_to_svg_colour(addr.weight().get_type()).to_hex(),
                    )
                    .set("stroke-width", "1")
            }),
            Self::CubicBezier { points, addr, .. } => Box::new({
                let data = Data::new()
                    .move_to((points[0].x, points[0].y))
                    .cubic_curve_to((
                        points[1].x,
                        points[1].y,
                        points[2].x,
                        points[2].y,
                        points[3].x,
                        points[3].y,
                    ));
                Path::new()
                    .set("d", data)
                    .set("fill", "none")
                    .set(
                        "stroke",
                        wire_type_to_svg_colour(addr.weight().get_type()).to_hex(),
                    )
                    .set("stroke-width", 1)
            }),
            Self::Arrow { .. } => {
                panic!("Arrows should not be in svgs")
            }
        }
    }
}

impl<T: Ctx> Shapes<T>
where
    <<T as Ctx>::Edge as WithWeight>::Weight: WithType,
{
    #[must_use]
    pub fn to_svg(&self) -> Document {
        let mut document = Document::new()
            .set("width", self.size.x * SCALE)
            .set("height", self.size.y * SCALE);

        let scale = TSTransform::from_scaling(SCALE);

        for shape in &self.shapes {
            let mut shape = shape.clone();
            shape.apply_tst_transform(&scale);
            document = document.add(shape.to_svg());
        }

        document
    }
}

fn wire_type_to_svg_colour(wire_type: WireType) -> Color32 {
    match wire_type {
        WireType::Data => Color32::BLACK,
        WireType::ControlFlow => Color32::GOLD,
        WireType::SymName => Color32::DARK_GREEN,
        WireType::Colour(colour) => colour,
    }
}
