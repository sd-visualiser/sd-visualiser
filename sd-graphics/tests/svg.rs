use dir_test::{Fixture, dir_test};
use sd_core::{
    language::sd_lang::Expr,
    lp::Solver,
    monoidal::{graph::MonoidalGraph, wired_graph::from_graph},
};
use sd_graphics::{layout::layout, render::generate_shapes, shape::Shapes};
use svg::Document;

fn to_svg(expr: &Expr) -> Document {
    let solver = Solver::default();
    let graph = expr.to_graph(false).unwrap();
    let monoidal_term = from_graph(&graph, solver);
    let monoidal_graph = MonoidalGraph::from(&monoidal_term);
    let layout = layout(&monoidal_graph, solver).unwrap();
    let mut shapes = Vec::new();
    generate_shapes(&mut shapes, &layout, true);
    let shapes = Shapes {
        shapes,
        size: layout.size(),
    };
    shapes.to_svg()
}

#[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "**/with*.sd", loader: sd_core::language::sd_lang::parse_sd_lang, postfix: "external_inputs")]
fn external_inputs(fixture: Fixture<(&str, Expr)>) {
    let (name, expr) = fixture.content();
    let svg = to_svg(expr);
    insta::assert_binary_snapshot!(
        &format!("external_inputs_{name}.svg"),
        svg.to_string().into(),
        &svg.to_string()
    );
}
