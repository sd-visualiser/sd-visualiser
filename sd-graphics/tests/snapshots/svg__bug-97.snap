---
source: sd-graphics/tests/svg.rs
expression: "<svg height=\"513.72986\" width=\"370\" xmlns=\"http://www.w3.org/2000/svg\">\n<rect fill=\"none\" height=\"463.72986\" stroke=\"gray\" stroke-width=\"1\" width=\"320\" x=\"25\" y=\"25\"/>\n<circle cx=\"245\" cy=\"488.72986\" fill=\"black\" r=\"2.5\"/>\n<circle cx=\"125\" cy=\"488.72986\" fill=\"black\" r=\"2.5\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"125\" x2=\"125\" y1=\"313.72983\" y2=\"349.18887\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"245\" x2=\"245\" y1=\"313.72983\" y2=\"313.72983\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"125\" x2=\"125\" y1=\"399.18887\" y2=\"488.72986\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"245\" x2=\"245\" y1=\"463.72986\" y2=\"488.72986\"/>\n<path d=\"M125,288.72983 C125,301.22983,125,301.22983,125,313.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M245,288.72983 C245,301.22983,245,301.22983,245,313.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<rect fill=\"none\" height=\"238.72983\" stroke=\"gray\" stroke-width=\"1\" width=\"270\" x=\"50\" y=\"50\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"185\" x2=\"185\" y1=\"100\" y2=\"100\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"185\" x2=\"185\" y1=\"150\" y2=\"150\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"125\" x2=\"125\" y1=\"213.72983\" y2=\"213.72983\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"245\" x2=\"245\" y1=\"213.72983\" y2=\"213.72983\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"125\" x2=\"125\" y1=\"263.72983\" y2=\"288.72983\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"245\" x2=\"245\" y1=\"263.72983\" y2=\"288.72983\"/>\n<path d=\"M185,75 C185,75,185,80,185,100\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<g>\n<rect fill=\"white\" height=\"20\" rx=\"10\" ry=\"10\" stroke=\"black\" stroke-width=\"1\" width=\"60\" x=\"155\" y=\"65\"/>\n<text dominant-baseline=\"middle\" font-family=\"monospace\" font-size=\"16\" text-anchor=\"middle\" x=\"185\" y=\"75\">\ni64 1\n</text>\n</g>\n<path d=\"M185,100 C185,120.00001,185,125,185,125\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M185,125 C185,125,185,130,185,150\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<g>\n<rect fill=\"white\" height=\"20\" rx=\"10\" ry=\"10\" stroke=\"black\" stroke-width=\"1\" width=\"240\" x=\"65\" y=\"115\"/>\n<text dominant-baseline=\"middle\" font-family=\"monospace\" font-size=\"16\" text-anchor=\"middle\" x=\"185\" y=\"125\">\n%1 = trunc i64 1 to i32\n</text>\n</g>\n<path d=\"M185,150 C185,170,185,175,185,175\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M185,175 C149,175,125,182.74597,125,213.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M185,175 C221,175,245,182.74597,245,213.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<circle cx=\"185\" cy=\"175\" fill=\"black\" r=\"5\"/>\n<path d=\"M125,213.72983 C125,233.72984,125,238.72983,125,238.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M125,238.72983 C125,238.72983,125,243.72983,125,263.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<g>\n<rect fill=\"white\" height=\"20\" rx=\"10\" ry=\"10\" stroke=\"black\" stroke-width=\"1\" width=\"110\" x=\"70\" y=\"228.72983\"/>\n<text dominant-baseline=\"middle\" font-family=\"monospace\" font-size=\"16\" text-anchor=\"middle\" x=\"125\" y=\"238.72983\">\nret i32 %1\n</text>\n</g>\n<path d=\"M245,213.72983 C245,233.72984,245,238.72983,245,238.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M245,238.72983 C245,238.72983,245,243.72983,245,263.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M125,349.18887 C125,369.18887,125,374.18887,125,374.18887\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M125,374.18887 C125,374.18887,125,379.18887,125,399.18887\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M245,313.72983 C245,326.22983,245,326.22983,245,338.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M245,438.72986 C245,451.22986,245,451.22986,245,463.72986\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<rect fill=\"none\" height=\"100.00003\" stroke=\"gray\" stroke-width=\"1\" width=\"140\" x=\"175\" y=\"338.72983\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"245\" x2=\"245\" y1=\"338.72983\" y2=\"363.72983\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"245\" x2=\"245\" y1=\"413.72986\" y2=\"438.72986\"/>\n<path d=\"M245,363.72983 C245,383.72983,245,388.72983,245,388.72983\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M245,388.72983 C245,388.72983,245,393.72986,245,413.72986\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<g>\n<rect fill=\"white\" height=\"20\" rx=\"10\" ry=\"10\" stroke=\"black\" stroke-width=\"1\" width=\"110\" x=\"190\" y=\"378.72983\"/>\n<text dominant-baseline=\"middle\" font-family=\"monospace\" font-size=\"16\" text-anchor=\"middle\" x=\"245\" y=\"388.72983\">\nret i32 %1\n</text>\n</g>\n</svg>"
extension: svg
snapshot_kind: binary
---
